### PR TITLE
tunnel: Fix goroutine leak

### PR DESCRIPTION
**Pull request by cg505**
The goroutine reading from the TCP might miss when the tunnel closed, and would
get stuck trying to write to a channel forever. If the tunnel connection was
dying a lot, this could cause a pretty substantial leak. This makes sure to
cancel the goroutine whenever we stop tunneling.

Unfortunately this leak is a little hard to simulate, so it might take a few days (of time, not work) to "verify" that the fix is working as expected. I think this is strictly more correct than what we do right now, so we can go ahead and merge it whenever though. We will also want to update the node-controller code to use the new pkg version when this is merged.

### CODE REVIEW

File: pkg/tunnel/tunnel.go [hunk 1]
```diff
@@ -71,7 +71,6 @@ func Client(scc node.ControllerClient, ln net.Listener, token,
 
 func connect(scc node.ControllerClient, stream net.Conn,
 	token, name string, port uint32) {
-	defer stream.Close()
```
Comment (kklin) [inline pkg/tunnel/tunnel.go]: Why remove this?

**Review by kklin — COMMENTED** [review]
Looks reasonable to me, although I don't know this code all that well.

I'll also leave it to you to decide how we should decide if we can close the bug (options range from just assuming this fixes it, to trying to recreate the bug)

File: pkg/tunnel/tunnel.go (hunk 1 above)
Comment (cg505) [reply pkg/tunnel/tunnel.go]: It's handled in `streamBidirectional`, but I realize now there are error cases that won't call that function, so we should probably leave it.

**Comment by cg505** [comment]
Let's merge this now, I will also test over a longer period of time this week.

**Comment by cg505** [comment]
I'm pretty confident it works at this point.

**Pull request merged**